File: storageManager/googleStorage.js

// Vendor Modules
/////////////////////////
// var async = require('async')
var _ = require('lodash')
var gcloud = require('gcloud')

var config =
{
	bucket: 'just_a_test'
}



// Our Modules
/////////////////////////
var StorageManager = require('./storageManager')

// Main Script
/////////////////////////
var googleStorage = module.exports = StorageManager.Base.extend({

start: function(options, callback)
{
	this.storage = gcloud.storage(
	{
		projectId: options.storage.google.projectId,
		keyFilename: options.storage.google.keyFilename,
	})
	this.bucket = this.storage.bucket(options.storage.google.bucket)
	callback()
},




listFiles: function(path, callback)
{
	this.bucket.getFiles(
	{
		prefix: path
	},
	function(err, files)
	{
		if (err)
		{
			return callback(err)
		}
		var fileData=[]
		_.forEach(files, function(entry)
		{
			var data =
			{
				key: entry.metadata.name,
				name: entry.metadata.name,
				size: entry.metadata.size,
				updated: entry.metadata.updated
			}
			fileData.push(data)
		})

		callback(null, fileData)
	})
},

getFile: function(sourcePath, destinationPath, callback)
{
	var fileToDownload = this.bucket.file(sourcePath)
	fileToDownload.download(
		{
			destination: destinationPath
		},
		callback)
},

addFile: function(sourcePath, destinationPath, callback)
{
	var options =
	{
		destination: destinationPath,
	//	resumable: true,
		validation: 'crc32c',
	}
	this.bucket.upload(sourcePath, options, callback)
},

// May not retrieve a valid link!
//If this were to be bercome a more secure private app, better to use file.getSignedUrl method
// which returns a temporary link (time to be specified)
getFileUrl: function(file, callback)
{
	var url = 'http://storage.googleapis.com/'+config.bucket+'/'+file
	callback(null, url)
},



deleteFile : function(file, callback)
{
	var fileToDelete = this.bucket.file(file)
	fileToDelete.delete(callback)
},



isFile: function(file, callback)
{
	var fileToGet = this.bucket.file(file)
	fileToGet.getMetadata(function(err)
			{
			 if (err)
				 callback(null, false)
			else
				callback(null, true)
			})
},

// end of class
})

StorageManager.addPlugin('google', googleStorage)

File: storageManager/googleStorage.py

from gcloud import storage
#import storageManager
import os


#GOOGLE_APPLICATION_CREDENTIALS = 'c:/temp/googlePythonKey.json'

os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = 'c:/temp/googlePythonKey.json' #<path to private key>
os.environ['GCLOUD_DATASET_ID'] = 'metal-sky-96603'

class GoogleStorage(object):

	def __init__(self, options):
		self.bucket = storage.get_bucket('just_a_test')
		self.bucket.make_public(recursive=True, future=True)

	def listFiles(self, path):
		return list(self.bucket.list_blobs())

	def getFile(self, sourcePath, destinationPath):
		try:
			blob = self.bucket.get_blob(sourcePath)
			blob.download_to_filename(destinationPath)
		except AttributeError:
			raise OSError('the File was not found on the google filesystem')

	def addFile(self, sourcePath, destinationPath):
		self.bucket.upload_file(sourcePath, destinationPath)

	def deleteFile(self, fileName):
		try:
			blob = self.bucket.get_blob(fileName)
			blob.delete()
		except AttributeError:
			raise OSError('The file to delete was not found on the Google Filesystem')

	def getFileUrl(self, fileName):

		blob = self.bucket.get_blob(fileName)
		if blob:
			result = 'https://www.googleapis.com/storage/'+self.bucket.id+'/'+fileName
			return result
		else:
			raise OSError('The file was not found on the google filesystem')


	def isFile(self, fileName):
		blob = self.bucket.get_blob(fileName)
		if blob:
			return True
		else:
			return False

